### PR TITLE
WIP: Issue #90:  Bundle uPortal-home with uPortal-start

### PR DESCRIPTION
**Pull request by apetro**
Work in progress.

##### Checklist

-   [x] the [individual contributor license agreement][] is signed
-   [x] commit message follows [commit guidelines][]
-   [ ] tests are included
-   [ ] documentation is changed or added
-   [ ] [message properties][] have been updated with new phrases
-   [ ] view conforms with [WCAG 2.0 AA][]

[individual contributor license agreement]: https://github.com/Jasig/uPortal/blob/master/CONTRIBUTING.md#individual-contributor-license-agreement
[commit guidelines]: https://github.com/Jasig/uPortal/blob/master/CONTRIBUTING.md#commit
[message properties]: https://github.com/Jasig/uPortal/tree/master/uportal-war/src/main/resources/properties/i18n
[WCAG 2.0 AA]: https://www.w3.org/WAI/WCAG20/quickref/?levels=aaa&technologies=smil%2Cpdf%2Cflash%2Csl

### CODE REVIEW

**Comment by bpowell** [comment]
Is there a version in maven central for uPortal-home?

**Comment by ChristianMurphy** [comment]
Two actually:

The Java+JavaScript official release: https://search.maven.org/#search%7Cga%7C1%7Ca%3A%22uportal-home%22

And the WIP JavaScript only unofficial release: https://search.maven.org/#search%7Cga%7C1%7Ca%3A%22uportal__home%22

**Comment by apetro** [comment]
Looks like the AppVeyor build failure is `[ant:java] ERROR [51:36.999] Failed: importing file [C:\projects\uportal-start\data\quickstart\portlet-definition\about-college-life.portlet-definition.xml]`

Import/Export generally and that entity file specifically aren't touched by this changeset, so likely a failure not-related-to-this-changeset.

**Comment by ChristianMurphy** [comment]
@apetro I believe that will be patched when https://github.com/Jasig/uPortal/pull/1066 is releases as part of version 5.0.3

**Comment by drewwills** [comment]
Oh this is a bummer...

```
FAILURE: Build failed with an exception.

* What went wrong:
Could not resolve all files for configuration ':overlays:KeyValueStore:runtimeClasspath'.
> Could not find edu.wisc.my:key-value-store:3.0.0.
  Searched in the following locations:
      file:/home/drew/.m2/repository/edu/wisc/my/key-value-store/3.0.0/key-value-store-3.0.0.pom
      file:/home/drew/.m2/repository/edu/wisc/my/key-value-store/3.0.0/key-value-store-3.0.0.war
      https://repo.maven.apache.org/maven2/edu/wisc/my/key-value-store/3.0.0/key-value-store-3.0.0.pom
      https://repo.maven.apache.org/maven2/edu/wisc/my/key-value-store/3.0.0/key-value-store-3.0.0.war
  Required by:
      project :overlays:KeyValueStore
```

@apetro @vertein -- how can we get a release of `key-value-store` in Maven Central?

**Comment by ChristianMurphy** [comment]
> @apetro @vertein -- how can we get a release of key-value-store in Maven Central?

related to https://github.com/UW-Madison-DoIT/KeyValueStore/issues/26

**Comment by vertein** [comment]
Key Value Store 3.1.1 is now in Maven Central

```
<groupId>edu.wisc.my.apps</groupId>
<artifactId>key-value-store</artifactId>
```

**Comment by drewwills** [comment]
@vertein -- will 3.1.1 work?  (@apetro advised me to use 3.0.0.)

**Comment by vertein** [comment]
@drewwills -- we aren't using version 3.1.1, but it should still work.  Only new features added, no breaking changes from 3.0.0.

File: overlays/storage/build.gradle [hunk 1]
```diff
@@ -0,0 +1,7 @@
+dependencies {
+    runtime "edu.wisc.my:key-value-store:${keyValueStoreVersion}@war"
```
Comment (bjagg) [inline overlays/storage/build.gradle]: edu.wisc.my.**apps**

Just noting the coordinate difference.

Comment (drewwills) [reply overlays/storage/build.gradle]: Coordinate?

**Comment by drewwills** [comment]
@apetro @vertein...

missing-font

**Comment by drewwills** [comment]
@apetro...

hacked-json-apis

WRT "Add /marketplace/entry/{fname}.json to get JSON for a single Marketplace entry" -- this one seems to be present in vanilla uPortal.

**Comment by bjagg** [comment]
Yes, some API customizations to support uPortal-home have been migrated

**Comment by drewwills** [comment]
Okay, as of this moment (5:30 Friday, MDT), if you...

  - Use this branch
  - Run `$ ./gradlew install` on uPortal `master` to put a copy of uPortal `3.1.0-SNAPSHOT` in your local Maven repo
  - Configure `uPortal-home.enabled=true` in `global.properties`
  - Run `$ ./gradlew portalInit tomcatStart` in uPortal-start

You will see...

uportal-home

Most everything seems to "work" in a basic sense;  support for widgets is missing.  (Widget support is pending metadata being converted to publishing parameters.)

**Comment by bjagg** [comment]
Is this ready to merge?

**Comment by drewwills** [comment]
I'd be very happy to see it merge soon, but we need to...

  - Set up the data we want to use with it (special tab, only for uPortal-home)
  - Address the issue of WCAG 2.0 Level AA

The first is easy, the second is a bit of a quandary.

**Comment by ChristianMurphy** [comment]
@davidmsibley @apetro @vertein How is WCAG AA progressing?

**Comment by apetro** [comment]
[MyUW's current roadmap](http://go.wisc.edu/myuw-roadmap) anticipates achieving WCAG A (not AA) before October 2018 and then AA before January 2019.

**Comment by ChristianMurphy** [comment]
@apetro what's the status of this PR and the progress toward WCAG AA?

**Comment by ChristianMurphy** [comment]
@apetro is this PR still something UW is actively working towards?

**Comment by ChristianMurphy** [comment]
@apetro @nogalpaulina has WCAG 2.0 AA accessibility on uP-home progressed?
Is there still interest in moving this forward and getting uP-home compatible with uP-start?

There is some renewed community interest in this https://github.com/Jasig/uPortal-start/issues/90#issuecomment-713139964

**Comment by remyjacob** [comment]
First of all, I'm pretty new to uPortal-start / Uportal-home and I'm not a java/maven/gradle/... specialist, so do not assume I can read between the lines when it comes to configuration tasks..

I tried to follow all configuration steps decribed above and found the following issues 

**1 - Issue #90: Point the 'Log In' button on uPortal-home to '/cas/login?service=http://localhost:8080/uPortal/Login' (hard-coded, but better than pointing to 'my.wisc.edu')** 

The path to `overlays/web/src/main/webapp/js/override.js` should be `overlays/web/web/src/main/webapp/js/override.js`

**2 - Issue #90: Add rendering pipeline configs for uPortal-home**

I first tried to manually copy `overlays/uPortal/src/main/resources/properties/contextOverrides/overridesContext.xml`  to `overlays/uPortal/src/main/webapp/WEB-INF/classes/properties/optionalContext` but as it still does not worked at runtime, I included uPortalHomeRenderingPipelineContext.xml  content into overridesContext.xml

**3 - Task :overlays:web:tomcatDeploy**

It fails with the following error "Cannot expand ZIP '`/opt/sources/uPortal-start/overlays/web/build/libs/web.war`' as it does not exist."
This file is in `/overlays/web/web/target/web.war`
So I had to `"cp overlays/web/web/target/web.war /opt/sources/uPortal-start/overlays/web/build/libs/web.war"`

**Comment by remyjacob** [comment]
I managed customizing, building and deploying uportal-home as an overlay to uPortal-start.
I also managed customizing, building and deploying uportal-app-framework as a standalone webapp.

But i would like to deploy all together using gradle.
Ho do i include customizations i've done in uportal-app-framework in the main uPortal-start deployment tasks ?

**Comment by apetro** [comment]
No, the MyUW team is not actively working on bundling uPortal-home with uPortal-start etc. I don't think MyUW is likely to pick that up again as a priority. In general MyUW paths forward may include strategies like re-implementing the valuable parts of uPortal-app-framework as Web Components ( https://github.com/myuw-web-components/ ) so that new front-end applications can be developed without dependence on AngularJS or on any other specific front-end framework. (Browser standards for the win).

**Comment by bjagg** [comment]
Is there still interest in uPortal-home integration, or can I close this?

**Comment by bjagg** [comment]
uPortal-home is being phased out.

**Pull request closed without merging**